lib/ansible/plugins/action/add_host.py

```diff
@@ -58,4 +60,26 @@ def run(self, tmp=None, task_vars=dict()):
 
         return dict(changed=True, add_host=dict(host_name=new_name, groups=new_groups, host_vars=host_vars))
 
+def _parse_ip_host_and_port(hostname):
+    """
+    Attempt to parse the hostname and port from a hostname, e.g.,
 
+    some-host-name
+    some-host-name:80
+    8.8.8.8
+    8.8.8.8:80
+    2001:db8:0:1
+    [2001:db8:0:1]:80
+    """
+    if hostname.count(':') > 1:
+        match = re.match(
+            '\[(?P<ip>[^\]]+)\](:(?P<port>[0-9]+))?',
+            hostname
+        )
+        if match:
+            return match.group('ip'), match.group('port')
+        else:
+            return hostname, None
+    elif ':' in hostname:
+        return hostname.split(':')
```

Comment:
I like to give split a second parameter to control the maximum number of times it will split.

So something like: ``` hostname.split(':', 1)```  it may also be appropriate to use rsplit() instead of split() here (although the ipv6 address might be the only case where rsplit() vs split() would be valid and you're already taking care of that above.)